VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,545 @@
+     */
+    mapping (string => EnumerableSetUpgradeable.UintSet) private _regionDataCenterIds;
+
+    /**
+     * @notice Additional information for `datacenterId`
+     * @dev Contains removed status & number of connected storage nodes
+     */
+    mapping (uint => DatacenterInfo) private _datacenterInfo;
+
+    /**
+     * @notice StorageNode by nodeId
+     */
+    mapping (uint => StorageNode) private _nodeMap;
+
+    /**
+     * @notice UnregisterTime of each storage node
+     * @dev Value is over 0 if unregistered
+     */
+    mapping (uint => uint) private _nodeUnregisterTime;
+
+    /**
+     * @notice nodeId per did address
+     */
+    mapping (address => uint) private _didNodeId;
+
+    /** 
+     * @notice nodeId per endpointUri
+     */
+    mapping (string => uint) private _endpointNodeId;
+
+    /**
+     * @notice nodeId list per country code
+     */
+    mapping (string => EnumerableSetUpgradeable.UintSet) private _countryNodeIds;
+
+    /**
+     * @notice nodeId list per region code
+     */
+    mapping (string => EnumerableSetUpgradeable.UintSet) private _regionNodeIds;
+
+
+    /**
+     * @notice datacenterId counter
+     * @dev starts from 1
+     */
+    CountersUpgradeable.Counter private _datacenterIdCounter;
+    /**
+     * @notice nodeId counter
+     * @dev starts from 1
+     */
+    CountersUpgradeable.Counter private _nodeIdCounter;
+
+    /**
+     * @notice Denominator for latitude & longitude values
+     */
+    uint8 public constant DECIMAL = 8;
+
+    /**
+     * @notice Additional information for a data center
+     * @dev Used internally inside the contract
+     * @param isActive True when added. False after removed
+     * @param nodeCount Number of connected storage nodes
+     */
+    struct DatacenterInfo {
+        bool isActive;
+        uint nodeCount;
+    }
+
+    error InvalidDatacenterName();
+    error InvalidCountryCode();
+    error InvalidRegionCode();
+    error InvalidLatitude();
+    error InvalidLongitude();
+    error InvalidDatacenterId(uint id);
+    error HasDependingNodes();
+    error InvalidEndpointUri();
+    error InvalidDIDAddress();
+    error InvalidUnregisterTime();
+
+    // constructor() {
+    //     _disableInitializers();
+    // }
+
+    /**
+     * @dev initializer of deployment
+     */
+    function initialize() initializer public {
+        __VDAVerificationContract_init();
+    }
+
+    /**
+     * @notice Check validity of country code
+     * @param countryCode Unique two-character string code
+     */
+    function validateCountryCode(string calldata countryCode) internal pure {
+        if (bytes(countryCode).length != 2) {
+            revert InvalidCountryCode();
+        }
+    }
+
+    /**
+     * @notice Check validity of region code
+     * @param regionCode Unique region string code
+     */
+    function validateRegionCode(string calldata regionCode) internal pure {
+        if (bytes(regionCode).length == 0) {
```

Comment:
Also check max-length based on values in https://github.com/verida/verida-js/blob/main/packages/account-node/src/countries.ts